### PR TITLE
Operator

### PR DESCRIPTION
**Pull request by VaibhavPage**

### CODE REVIEW

File: common/common.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package common
+
```
Comment (artbegolli) [inline common/common.go]: Cool - going forward we can use this file for any other constants

File: controllers/config.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+		&corev1.ConfigMap{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if cm, ok := obj.(*corev1.ConfigMap); ok {
+					ctrl.logger.Info("detected configmap update. updating the controller config")
+					err := ctrl.updateConfig(cm)
+					if err != nil {
+						ctrl.logger.Error("update of config failed", "err", err)
+					}
```
Comment (artbegolli) [inline controllers/config.go]:
```go
if err := ctrl.updateConfig(cm); err != nil {
    ctrl.logger.WithError(err).Errorln("update of config failed")
}
```

Can we remove the unnecessary line and update the log to use `WithError`

File: controllers/config.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+				}
+			},
+			UpdateFunc: func(old, new interface{}) {
+				if newCm, ok := new.(*corev1.ConfigMap); ok {
+					ctrl.logger.Info("detected configmap update. updating the controller config.")
+					err := ctrl.updateConfig(newCm)
+					if err != nil {
+						ctrl.logger.WithError(err).Error("update of config failed")
+					}
```
Comment (artbegolli) [inline controllers/config.go]: Same as above, can be reduced. Also, if we receive an error here we should be returning

File: controllers/config.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+			},
+		})
+
+	go controller.Run(ctx.Done())
+	return controller, nil
+}
+
+// creates a new config map watcher
+func (ctrl *Controller) newControllerConfigMapWatch() *cache.ListWatch {
+	x := ctrl.kubeClient.CoreV1().RESTClient()
```
Comment (artbegolli) [inline controllers/config.go]: Can we rename this to something more useful

File: controllers/config.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	return &cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// ResyncConfig reloads the controller config from the configmap
+func (ctrl *Controller) ResyncConfig(namespace string) error {
+	cmClient := ctrl.kubeClient.CoreV1().ConfigMaps(namespace)
+	cm, err := cmClient.Get(ctrl.Configmap, metav1.GetOptions{})
+	if err != nil {
+		return err
```
Comment (artbegolli) [inline controllers/config.go]: Can we add an error log here

File: controllers/config.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+// updates the controller configuration
+func (ctrl *Controller) updateConfig(cm *corev1.ConfigMap) error {
+	configStr, ok := cm.Data[common.ControllerConfigMapKey]
+	if !ok {
+		return fmt.Errorf("configMap '%s' does not have key '%s'", ctrl.Configmap, common.ControllerConfigMapKey)
+	}
+	var config *ControllerConfig
+	err := yaml.Unmarshal([]byte(configStr), &config)
+	if err != nil {
+		return err
```
Comment (artbegolli) [inline controllers/config.go]: Can we add an error log here

File: controllers/controller.go [hunk 7]
```diff
@@ -15,3 +15,173 @@ limitations under the License.
 */
```
Comment (artbegolli) [inline controllers/controller.go]: Let's remove all the `fmt.Printf` here and use a logger instead

File: controllers/controller.go [hunk 8]
```diff
@@ -15,3 +15,173 @@ limitations under the License.
...
+
+	// due to the base delay of 5ms of the DefaultControllerRateLimiter
+	// requeues will happen very quickly even after a ocibuilder pod goes down
+	// we want to give the ocibuilder pod a chance to come back up so we give a generous number of retries
+	if ctrl.queue.NumRequeues(key) < 20 {
+		// Re-enqueue the key rate limited. This key will be processed later again.
+		ctrl.queue.AddRateLimited(key)
+		return nil
+	}
+	return errors.New("exceeded max requeues")
```
Comment (artbegolli) [inline controllers/controller.go]: I think we should log the original error here as well to give the user some context on what has failed

File: controllers/controller.go [hunk 9]
```diff
@@ -15,3 +15,173 @@ limitations under the License.
...
+	if !ok {
+		fmt.Printf("key '%s' in index is not a builder", key)
+		return true
+	}
+
+	ctx := newOperationContext(builder, ctrl)
+
+	err = ctx.operate()
+	if err != nil {
+		ctrl.logger.WithError(err).WithField(common.LabelOCIBuilderName, builder.Name).Errorln("failed to operate on the ocibuilder obejct")
```
Comment (artbegolli) [inline controllers/controller.go]: Should we be returning false here?

File: version.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (artbegolli) [inline version.go]: Versioning logic is already available in https://github.com/ocibuilder/ocibuilder/blob/master/provenance/provenance.go

Comment (artbegolli) [reply version.go]: We should append anything that you see missing there

File: controllers/controller.go (hunk 9 above)
Comment (VaibhavPage) [reply controllers/controller.go]: No, you wan't to re-queue the key

File: controllers/controller.go (hunk 8 above)
Comment (VaibhavPage) [reply controllers/controller.go]: It is already logged before calling handleErr.

File: version.go (hunk 10 above)
Comment (VaibhavPage) [reply version.go]: Cool, didn't know it was there

File: controllers/config.go (hunk 5 above)
Comment (VaibhavPage) [reply controllers/config.go]: handled by caller.

File: controllers/config.go (hunk 6 above)
Comment (VaibhavPage) [reply controllers/config.go]: handled by caller

File: controllers/operator.go [hunk 11]
```diff
@@ -15,3 +15,37 @@ limitations under the License.
...
+	"github.com/sirupsen/logrus"
+)
+
+// the context of an operation on a ocibuilder object.
+// the controller creates this context each time it picks a ocibuilder object off its queue.
+type operationContext struct {
+	// builder is the ocibuilder object
+	builder *v1alpha1.OCIBuilder
+	// updated indicates whether the ocibuilder object was updated and needs to be persisted back to k8
+	updated bool
```
Comment (artbegolli) [inline controllers/operator.go]: This is giving a linting error for being unused

Comment (VaibhavPage) [reply controllers/operator.go]: hmm, its gonna get used later in the code. I can comment it for now.

Comment (artbegolli) [reply controllers/operator.go]: cool

Comment (VaibhavPage) [reply controllers/operator.go]: let me check if updated: false in init struct satisfies lint.

Comment (VaibhavPage) [reply controllers/operator.go]: I don't like that the CI lint didnt understand the default value for bool and complained unnecessarily.

**Comment by artbegolli** [comment]
Linting is being killed because of a bug in golangci-lint. This is fixed in version v1.21.0

Try bumping the version in config.yml from v1.19.0 to v1.21.0
`curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b $(go env GOPATH)/bin v1.21.0`

**Comment by artbegolli** [comment]
> Linting is being killed because of a bug in golangci-lint. This is fixed in version v1.21.0
> 
> Try bumping the version in config.yml from v1.19.0 to v1.21.0
> `curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b $(go env GOPATH)/bin v1.21.0`

pipeline looks good now

**Pull request merged**